src/EventStore.Core/TransactionLog/Scavenging/Stages/ChunkExecutor.cs

```diff
@@ -0,0 +1,375 @@
+			ITFChunkScavengerLog scavengerLogger,
+			CancellationToken cancellationToken) {
+
+			Log.Trace("SCAVENGING: Starting new scavenge chunk execution phase for {scavengePoint}",
+				scavengePoint.GetName());
+
+			var checkpoint = new ScavengeCheckpoint.ExecutingChunks(
+				scavengePoint: scavengePoint,
+				doneLogicalChunkNumber: default);
+			state.SetCheckpoint(checkpoint);
+			Execute(checkpoint, state, scavengerLogger, cancellationToken);
+		}
+
+		public void Execute(
+			ScavengeCheckpoint.ExecutingChunks checkpoint,
+			IScavengeStateForChunkExecutor<TStreamId> state,
+			ITFChunkScavengerLog scavengerLogger,
+			CancellationToken cancellationToken) {
+
+			Log.Trace("SCAVENGING: Executing chunks from checkpoint: {checkpoint}", checkpoint);
+
+			var startFromChunk = checkpoint?.DoneLogicalChunkNumber + 1 ?? 0;
+			var scavengePoint = checkpoint.ScavengePoint;
+
+			var physicalChunks = GetAllPhysicalChunks(startFromChunk, scavengePoint);
+
+			var borrowedStates = new IScavengeStateForChunkExecutorWorker<TStreamId>[_threads];
+			var stopwatches = new Stopwatch[_threads];
+
+			for (var i = 0; i < borrowedStates.Length; i++) {
+				borrowedStates[i] = state.BorrowStateForWorker();
+				stopwatches[i] = new Stopwatch();
+			}
+
+			try {
+				ParallelLoop.RunWithTrailingCheckpoint(
+					source: physicalChunks,
+					degreeOfParallelism: _threads,
+					getCheckpointInclusive: physicalChunk => physicalChunk.ChunkEndNumber,
+					getCheckpointExclusive: physicalChunk => {
+						if (physicalChunk.ChunkStartNumber == 0)
+							return null;
+						return physicalChunk.ChunkStartNumber - 1;
+					},
+					process: (slot, physicalChunk) => {
+						// this is called on other threads
+						var concurrentState = borrowedStates[slot];
+						var sw = stopwatches[slot];
+
+						// the physical chunks do not overlap in chunk range, so we can sum
+						// and reset them concurrently
+						var physicalWeight = concurrentState.SumChunkWeights(
+								physicalChunk.ChunkStartNumber,
+								physicalChunk.ChunkEndNumber);
+
+						if (physicalWeight > scavengePoint.Threshold || _unsafeIgnoreHardDeletes) {
+							ExecutePhysicalChunk(
+								physicalWeight,
+								scavengePoint,
+								concurrentState,
+								scavengerLogger,
+								physicalChunk,
+								sw,
+								cancellationToken);
+
+							// resetting must happen after execution, but need not be in a transaction
+							// which is handy, because we cant run transactions concurrently very well
+							// https://www.sqlite.org/cgi/src/doc/begin-concurrent/doc/begin_concurrent.md)
+							concurrentState.ResetChunkWeights(
+								physicalChunk.ChunkStartNumber,
+								physicalChunk.ChunkEndNumber);
+						} else {
+							Log.Trace(
+								"SCAVENGING: skipped physical chunk: {oldChunkName} " +
+								"with weight {physicalWeight:N0}. ",
+								physicalChunk.Name,
+								physicalWeight);
+						}
+						cancellationToken.ThrowIfCancellationRequested();
```

Comment:
is this exception actually caught somewhere? otherwise could it result in a global unhandled exception? (causing the application to exit)

Reply:
yep it ends up back on the thread that called the loop. it's covered by this test `exception_during_processing_is_propagated`

Reply:
awesome, thanks